### PR TITLE
Require c++17

### PR DESCRIPTION
**Pull request by havogt**
Minimal requirements updated:
- CMake 3.18.1 (CMake 3.18 is the first version to support C++17 for CUDA compilers)
- CUDA 11.0 (first version to support C++17)

The following is updated/removed as it is no longer needed:
- removed CMake workaround for Threads
- removed custom type_traits.hpp and numeric.hpp as (meta-) functions are in c++17
- no special macro to disable constexpr for NVCC
- removed some pre-c++17 code that had a better c++17 version
- removed workarounds for CUDA 10.x

CI:
- updated compilers, see README.md
- especially: daint `nvcc_clang` environment (which was clang 7) is replaced by `nvcc_cray` (cray 10 + cuda 11.2)

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

File: include/gridtools/reduction/gpu.hpp [hunk 1]
```diff
@@ -18,7 +18,6 @@
 #include "../common/cuda_util.hpp"
 #include "../common/host_device.hpp"
 #include "../common/integral_constant.hpp"
-#include "../common/numeric.hpp"
```
Comment (anstaf) [inline include/gridtools/reduction/gpu.hpp:21]: please add
```c++ 
#include <numeric> 
```

File: include/gridtools/stencil/core/compute_extents_metafunctions.hpp [hunk 2]
```diff
@@ -38,7 +38,7 @@ namespace gridtools {
                 template <intent Intent>
                 struct has_intent {
                     template <class Item, class Param = meta::second<Item>>
-                    using apply = bool_constant<Param::intent_v == Intent>;
+                    using apply = std::bool_constant<Param::intent_v == Intent>;
```
Comment (anstaf) [inline include/gridtools/stencil/core/compute_extents_metafunctions.hpp:41]: add `#include <type_traits>`

File: include/gridtools/stencil/core/convert_fe_to_be_spec.hpp [hunk 3]
```diff
@@ -56,7 +56,7 @@ namespace gridtools {
                         typename is_tmp_arg<Plh>::type,
                         typename get_data_type<Plh, DataStores>::type,
                         typename get_num_colors<Plh>::type,
-                        bool_constant<Accessor::intent_v == intent::in>,
+                        std::bool_constant<Accessor::intent_v == intent::in>,
```
Comment (anstaf) [inline include/gridtools/stencil/core/convert_fe_to_be_spec.hpp:59]: add `#include <type_traits>`

File: include/gridtools/stencil/core/esf_metafunctions.hpp [hunk 4]
```diff
@@ -37,7 +37,7 @@ namespace gridtools {
                 template <intent Intent>
                 struct has_intent {
                     template <class Item, class Param = meta::second<Item>>
-                    using apply = bool_constant<Param::intent_v == Intent>;
+                    using apply = std::bool_constant<Param::intent_v == Intent>;
```
Comment (anstaf) [inline include/gridtools/stencil/core/esf_metafunctions.hpp:40]: add include

File: include/gridtools/stencil/core/need_sync.hpp [hunk 5]
```diff
@@ -20,9 +20,9 @@ namespace gridtools {
                 template <class DirtyPlhs>
                 struct is_dirty_f {
                     template <class Item, class Extent = typename meta::second<Item>::extent_t>
-                    using apply = bool_constant<meta::st_contains<DirtyPlhs, meta::first<Item>>::value &&
-                                                (Extent::iminus::value != 0 || Extent::iplus::value != 0 ||
-                                                    Extent::jminus::value != 0 || Extent::jplus::value != 0)>;
+                    using apply = std::bool_constant<meta::st_contains<DirtyPlhs, meta::first<Item>>::value &&
```
Comment (anstaf) [inline include/gridtools/stencil/core/need_sync.hpp:23]: add include

File: include/gridtools/stencil/cpu_ifirst/entry_point.hpp [hunk 6]
```diff
@@ -41,8 +41,9 @@ namespace gridtools {
                         meta::transform<be_api::get_execution, stages_t>>::type;
                     using enclosing_extent_t = meta::rename<enclosing_extent,
                         meta::transform<be_api::get_extent, typename stages_t::plh_map_t>>;
-                    using fuse_all_t = bool_constant<all_parrallel_t::value && enclosing_extent_t::kminus::value == 0 &&
-                                                     enclosing_extent_t::kplus::value == 0>;
+                    using fuse_all_t =
+                        std::bool_constant<all_parrallel_t::value && enclosing_extent_t::kminus::value == 0 &&
```
Comment (anstaf) [inline include/gridtools/stencil/cpu_ifirst/entry_point.hpp:45]: add include

File: include/gridtools/storage/adapter/fortran_array_adapter.hpp [hunk 7]
```diff
@@ -66,7 +66,7 @@ namespace gridtools {
 
         using bindgen_view_rank = std::integral_constant<size_t, data_store_t::layout_t::unmasked_length>;
         using bindgen_view_element_type = std::remove_pointer_t<data_ptr_t>;
-        using bindgen_is_acc_present = bool_constant<true>;
+        using bindgen_is_acc_present = std::bool_constant<true>;
```
Comment (anstaf) [inline include/gridtools/storage/adapter/fortran_array_adapter.hpp:69]: add include; use `std::true_type`

File: include/gridtools/storage/data_store.hpp [hunk 8]
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include <memory>
+#include <numeric>
```
Comment (anstaf) [inline include/gridtools/storage/data_store.hpp:13]: is it used here?

File: include/gridtools/storage/sid.hpp [hunk 9]
```diff
@@ -56,7 +56,7 @@ namespace gridtools {
             auto filter_unmasked_bounds(layout_map<Is...>, Bounds const &bounds) {
                 using all_keys_t = get_keys<Bounds>;
                 using all_values_t = tuple_util::traits::to_types<Bounds>;
-                using is_unmasked_t = meta::list<bool_constant<Is >= 0>...>;
+                using is_unmasked_t = meta::list<std::bool_constant<Is >= 0>...>;
```
Comment (anstaf) [inline include/gridtools/storage/sid.hpp:59]: add include

File: tests/unit_tests/stencil/frontend/cartesian/test_call_stress_types.cpp [hunk 10]
```diff
@@ -7,7 +7,7 @@
  * Please, refer to the LICENSE file in the root directory.
  * SPDX-License-Identifier: BSD-3-Clause
  */
-#include <gridtools/meta/type_traits.hpp>
+
```
Comment (anstaf) [inline tests/unit_tests/stencil/frontend/cartesian/test_call_stress_types.cpp:10]: add `#include <type_traits>`

File: tests/unit_tests/test_meta.cpp [hunk 11]
```diff
@@ -260,14 +260,14 @@ namespace gridtools {
                           f<void, void, void, int, double, void, void>>::value,
             "");
 
-        // void_t (CWG 1558 https://wg21.cmeerw.net/cwg/issue1558)
+        // std::void_t (CWG 1558 https://wg21.cmeerw.net/cwg/issue1558)
         namespace defect_cwg_1558 {
```
Comment (anstaf) [inline tests/unit_tests/test_meta.cpp:264]: please remove the whole namespace -- it doesn't test any `meta` functionality now

**Review by anstaf — CHANGES_REQUESTED** [review]

**Comment by havogt** [comment]
launch jenkins

File: include/gridtools/storage/data_store.hpp (hunk 8 above)
Comment (anstaf) [reply include/gridtools/storage/data_store.hpp:13]: I checked the whole file -- `<numeric>` is not used, but `<utility>` is used (`std::move`).

**Review by anstaf — CHANGES_REQUESTED** [review]

File: include/gridtools/storage/data_store.hpp (hunk 8 above)
Comment (havogt) [reply include/gridtools/storage/data_store.hpp:13]: ups, I fixed it but didn't save the editor

**Review by anstaf — APPROVED** [review]

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch mpitests

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
Jenkins failures are due to the dropped `nvcc_clang` environment.

**Pull request merged**